projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/spm/spm_dlsym.cpp

```diff
@@ -20,50 +20,39 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-#pragma once
+#include "lib/rocprofiler-sdk/spm/spm_dlsym.hpp"
 
-#include <rocprofiler-sdk/defines.h>
-#include <rocprofiler-sdk/fwd.h>
+#include <dlfcn.h>
+#include <atomic>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <vector>
 
-ROCPROFILER_EXTERN_C_INIT
-
-/**
- * @defgroup SPM_SERVICE SPM Service
- * @brief Streaming Performance Monitoring
- *
- * @{
- */
-
-/**
- * @brief (experimental) ROCProfiler SPM Record.
- *
- */
-typedef struct ROCPROFILER_SDK_EXPERIMENTAL rocprofiler_spm_record_t
+namespace rocprofiler
 {
-    /**
-     * Counters, including identifiers to get counter information and Counters
-     * values
-     */
-    rocprofiler_counter_record_t* counters;
-    uint64_t                      counters_count;
-} rocprofiler_spm_record_t;
-
-/**
- * @brief Configure SPM Service.
- *
- * @param [in] context_id
- * @param [in] buffer_id
- * @param [in] counter_config
- * @param [in] interval
- * @return ::rocprofiler_status_t
- */
-ROCPROFILER_SDK_EXPERIMENTAL
-rocprofiler_status_t
-rocprofiler_configure_spm_service(rocprofiler_context_id_t        context_id,
-                                  rocprofiler_buffer_id_t         buffer_id,
-                                  rocprofiler_counter_config_id_t counter_config,
-                                  uint64_t                        interval) ROCPROFILER_API;
-
-/** @} */
-
-ROCPROFILER_EXTERN_C_FINI
+namespace SPM
+{
+Dlsym::Dlsym()
+{
+    handle = dlopen("libhsa-amd-aqlprofile64.so", RTLD_NOW | RTLD_LOCAL | RTLD_NOLOAD);
+    if(!handle)
+        handle = dlopen("libhsa-amd-aqlprofile64.so.1", RTLD_NOW | RTLD_LOCAL | RTLD_NOLOAD);
+
+    if(!handle) return;
+
+    create_packets_fn = (CreateFn*) dlsym(handle, "aqlprofile_spm_create_packets");
+    delete_packets_fn = (DeleteFn*) dlsym(handle, "aqlprofile_spm_delete_packets");
+    spm_start_fn      = (StartFn*) dlsym(handle, "aqlprofile_spm_start");
+    spm_stop_fn       = (StopFn*) dlsym(handle, "aqlprofile_spm_stop");
+    spm_decode_fn     = (DecodeFn*) dlsym(handle, "aqlprofile_spm_decode_stream_v1");
+    spm_query_fn      = (QueryFn*) dlsym(handle, "aqlprofile_spm_decode_query");
+    is_supported_fn   = (SupportFn*) dlsym(handle, "aqlprofile_spm_is_event_supported");
+}
```

Comment:
You need a `ROCP_INFO` statement here reporting whether the dlsym is valid or not.